design/plugin-versioning.md

```diff
@@ -0,0 +1,232 @@
+# Plugin Versioning
+
+## Abstract
+This proposal outlines an approach to support versioning of Velero's plugin APIs to enable changes to those APIs to be made that are backwards compatible through the addition of new plugin methods.
+It will not support backwards incompatible changes such as method removal or method signature changes.
+
+## Background
+When changes are made to Velero’s plugin APIs, there is no mechanism for the Velero server to communicate the version of the API that is supported, or for plugins to communicate what version they implement.
+This means that any modification to a plugin API is a backwards incompatible change as it requires all plugins which implement the API to update and implement the new method.
+
+There are several components involved to use plugins within Velero.
+From the perspective of the core Velero codebase, all plugin kinds (e.g. ObjectStore, BackupItemAction) are defined by a single API interface and all interactions with plugins are managed by a plugin manager which provides an implementation of the plugin API interface for Velero to use.
+
+Velero communicates with plugins via gRPC.
+The core Velero project provides a framework (using the go-plugin project) for plugin authors to use to implement their plugins which manages the creation of gRPC servers and clients.
+Velero plugins import the Velero plugin library in order to use this framework.
+When a change is made to a plugin API, it needs to be made to the Go interface used by the Velero codebase, and also to the rpc service definition which is compiled to form part of the framework.
+As each plugin kind is defined by a single interface, when a plugin imports the latest version of the Velero framework, it will need to implement the new APIs in order to build and run successfully.
+If a plugin does not use the latest version of the framework, and is used with a newer version of Velero that expects the plugin to implement those methods, this will result in a runtime error as the plugin is incompatible.
+
+With this proposal, we aim to break this coupling and introduce plugin API versions.
+
+## Scenarios to Support
+The following describes interactions between Velero and its plugins that will be supported with the implementation of this proposal.
+For the purposes of this list, we will refer to existing Velero and plugin versions as `v1` and all following versions as version `n`.
+
+Velero client communicating with plugins:
+
+- Version `n` Velero client will be able to communicate with Version `n` plugin
+- Version `n` Velero client will be able to communicate with all previous versions of the plugin (Version `n-1` back to `v1`)
+- Optional: Previous versions of the Velero client (from version `v1` to version `n-1`) will be able to communicate with Version `n` plugin
+    - This may work but is not a requirement
+
+Plugin client communicating with other plugins:
```

Comment:
How do we differentiate "velero client" and "plugin client"?

Reply:
There isn't really a difference. I'm not aware of any plugins that currently call through to other plugins, but it was suggested that it could be a possibility in the future so I included it explicitly here. 

The resulting scenarios are the same though whether it's Velero or another plugin so I've just merged them.